File: Task4/src/Equals.scala
/**
 * Created by Шолохов on 02.02.2015.
 */
case class Equals (override val left: Expression, override val right: Expression) extends BinOp(left, right){
  override def apply(l : Boolean, r : Boolean) : Boolean = l == r
  override def toString(): String = "(" + left + "=" + right + ")"
  override def equals(o: Any): Boolean = {
    o match {
      case e : Equals => e.left.equals(left) && e.right.equals(right)
      case _ => false
    }
  }
}

File: Task4/src/Equals.java
import java.util.HashMap;

public class Equals extends Statement {
    public final Expression left;
    public final Expression right;

    public Equals(Expression left, Expression right) {
        this.left = left;
        this.right = right;
    }

    protected boolean compareImpl(String[] patterns, Statement other) {
        if (other instanceof Equals) {
            Equals eOther = (Equals) other;
            return left.equals(eOther.left) && right.equals(eOther.right);
        } else {
            return false;
        }
    }

    public Statement substituteTerm(Expression where, Expression what) {
        return new Equals(left.substituteTerm(where, what), right.substituteTerm(where, what));
    }

    public Statement substitutePatterns(Statement[] to) {
        return this;
    }

    public boolean estimate(HashMap<String, Boolean> values) {
        return false;
    }

    public String toString() {
        return "(" + left + " = " + right + ")";
    }
}
